### ISSUE TITLE
add epoch start time to distinguish between cached and active entries

### ISSUE THREAD

**Issue by colinbendell** [w3c/server-timing#43]
The UA may receive `Server-Timing` entries that were cached (eg: by a CDN), provide ancestral detail but may not be relevant to visualizing the current timeline from the UA's perspective. To allow the UA to distinguish between recent and historical `Server-Timing` values, servers should add a `start=` _epoch_ timestamp. This way the UA can distinguish between recent `Server-Timing` entries and past entries when visualizing the timeline.

# example 
For example, the following request: 
`curl -D - -o /dev/null -s https://colinbendell.cloudinary.com/image/upload/c_crop,f_auto,g_auto,h_350,w_400/v2/Wizard-Clap-by-Markus-Magnusson.gif` 

Consider the two responses:
```
date: Wed, 13 Dec 2017 04:04:39 GMT
server: cloudinary
content-type: image/gif
content-length: 580970

server-timing: cloudinary;dur=55;start=1513137879.239;total=55;desc="id_1a53021"
server-timing: cdn;dur=59;start=1513137879;total=114;desc="miss+rtt_12+id_4b6d04f2"
server-timing: cdn;dur=48;start=1513137879;total=162;desc="miss+rtt_36+id_b8c5515"
```

Here we have UA -> CDN -> CDN -> Server and all layers are active in this response which populates the CDN layers with a cach miss. 

A few minutes later we re-request the resource and we get:

```
date: Wed, 13 Dec 2017 04:25:29 GMT
server: cloudinary
server-timing: cloudinary;dur=55;start=1513137879.239;total=55
content-type: image/gif
content-length: 580970
server-timing: cdn;dur=1;start=1513139129;total=1;desc="hit+rtt_0+id_4b88b4b8"
server-timing: cdn;dur=13;start=1513139129;total=14;desc="hit+rtt_12+id_aff4e2c"
server-timing: cdn;dur=36;start=1513139129;total=50;desc="miss+rtt_33+id_84b8be06"
```

now we have UA -> CDN -> CDN -> CDN. 

In the second example, the UA had a total TTFB of 116ms. 

Using `Server-Timing` the UA is now able to distinguish between network time and server time. However, if the UA simply summed all the `Server-Timing` `dur=` values it would yield a total of 105ms as server think time. In reality 55ms of that 105ms total should not apply to visualizing this request. 

To solve this, the UA should look at the `start` epoch time and ignore entries that happen before the requested start time.

# Can we trust epoch time?

In this scenario, we can trust epoch start time because we are only worried about approximation to the request time to the UA as a mechanism to ignore older entries.

# Why not mark the 'last' `server-timing` entry as *last*?

I'm assuming the principle that each layer in the chain should have _isolated awareness_. that is, a request should only be responsible for adding itself and a) should never touch another server-timing entry and b) should never need to be aware of other server-timing entries that may or may not be present. In short, each layer should be ignorant of any other timing metrics in the eco system and therefore oblivious if it is the last or first hop in the chain.

# Shouldn't only the current server-timing entries be sent?
Just as the principle of _isolated awareness_, each hop should not be required to expunge past entries (though this may be desirable for business reasons). The ancestral details that Server-Timing provide are insightful for other purposes - debugging, triaging, utilization analytics, etc. For example, the Server-Timing values in the description may describe the A:B state from the origin and useful when combining with an analytics engine. The preservation of historical server-timing should be a business decision and not the responsibility of the spec.

**Comment by hauleth** [comment]
I would use `at` keyword for that and leave `start` to denote the "start time" within processed event itself. That would allow `server-timing` to be used to display real trace of the events that happened on the server.

**Comment by colinbendell** [comment]
Are you suggesting this would allow you to group `at` on a process layer with different `start` for each inner-process? Can you give a proposed example?

**Comment by hauleth** [comment]
@colinbendell sorry for late response but the idea is to use `start` keyword as a millisecond/nanosecond since imaginary start of the request handling by the server, so for example with headers like:

```
server-timing: request;dur=40;start=0
server-timing: db;dur=10;start=10
server-timing: json;dur=5;start=35
```

Would result in diagram like:

```
|--------------------------------------| request
          |--------| db
                                   |---| json
```

And `start` would be by default set to `0` (however no timing would be required to have `start=0`, so in theory all can start after starting handling request).

**Comment by noamr** [comment]
Since server timing exposes timings that come from different clocks so not 100% comparable, and OTOH considering that this data could be useful for [DevTools et al](https://bugs.chromium.org/p/chromium/issues/detail?id=594800), I believe perhaps a reasonable way to look at it is that one of the attributes would provide a start time *approximation*. I believe this would also solve #44.

Developers that want to look at it or at waterfalls created based on this data should be aware of the possible clock differences and that any waterfall is an approximation, though in some cases still useful.

Perhaps approx_start=${ISO-8601}?

/cc @yoavweiss @paulirish

**Comment by murraylchapman** [comment]
Lack of a standard for "start time" means that different vendors are implementing their own incompatible solutions:

Fastly [is going with](https://www.fastly.com/blog/supercharging-server-timing-http-trailers):

`Server-Timing: time-start-msec;dur=1544707555517,time-elapsed;dur=81`

Cloudinary [is going with](https://cloudinary.com/blog/inside_the_black_box_with_server_timing#what_details_are_you_sharing_):

`server-timing: cloudinary;dur=94;start=1516653847.986;total=94`

**Comment by hauleth** [comment]
One of the biggest problem I see there is that there is no API for fetching extra metadata from the entry (see #72). This would allow digesters to circumvent this problem in more uniform way.

**Comment by paulirish** [comment]
A month ago Cloudflare [started using a new convention](https://blog.cloudflare.com/new-standards/#measuring-impact:~:text=Server%2DTiming%20response%20headers): 

`server-timing: cfL4;desc="?proto=TCP&rtt=7337&sent=8&recv=8&lost=0&retrans=0&sent_bytes=3419&recv_bytes=832&delivery_rate=548023&cwnd=25&unsent_bytes=0&cid=94dae6b578f91145&ts=225`

**Comment by LPardue** [comment]
Hi @paulirish. I added `cfL4`. Mainly as a way to tightly couple with other RUM data. Other header fields aren't as accessible from the client-side javascript we have running.

We are shoehorning a lot into the `desc` but on balance it seemed like the most effective way to accommodate data values of arbitrary type or units. At least in the absence of something like #72. It's in early deployment but already proving useful to divide up our cohorts into groups based on things like RTT or packet loss.

Not sure exactly what cfL4 has to do with the OP issue of start time. Am I missing something?

Happy to answer any questions on our approach, or entertain other ideas.

**Comment by paulirish** [comment]
@LPardue Yup! Sorry I wasn't casting aspersions. :)  The solution seems clever. All good.

I've returned to this thread a few times to look at how various teams have created conventions for including a start time. So, I was adding mention to be comprehensive. Admittedly, it's a tad offtopic from Colin's original question.
